core/src/banking_stage/qos_service.rs

```diff
@@ -186,19 +190,22 @@ impl QosService {
         bank: &Bank,
     ) {
         let mut cost_tracker = bank.write_cost_tracker().unwrap();
+        let mut num_included = 0;
         transaction_cost_results
             .zip(transaction_committed_status)
             .for_each(|(tx_cost, transaction_committed_details)| {
                 // Only transactions that the qos service included have to be
                 // checked for update
                 if let Ok(tx_cost) = tx_cost {
+                    num_included += 1;
                     if let CommitTransactionDetails::Committed { compute_units } =
                         transaction_committed_details
                     {
                         cost_tracker.update_execution_cost(tx_cost, *compute_units)
                     }
                 }
             });
+        cost_tracker.sub_transactions_in_flight(num_included);
     }
 
     fn remove_transaction_costs<'a>(
```

Comment:
should sub_transactions_in_flight() here too

Reply:
I put this in wrong function, we should **only** do it in the `remove` costs calls

Reply:
Double checking: `count_try_added == count_remove_uncommitted + count_update_committed + count_remove`,  as I see from sce validator log: 
```
$ cat solana-validator.log | grep 'slot 255659546' | grep 'add tx cost success' | wc -l
1937
$ cat solana-validator.log | grep 'slot 255659546' | grep 'update committed tx cost' | wc -l
1917
$ cat solana-validator.log | grep 'slot 255659546' | grep 'remove tx cost' | wc -l
17
$ cat solana-validator.log | grep 'slot 255659546' | grep 'remove uncommitted tx cost' | wc -l
3
```
Or perhaps you count differently?

Reply:
I probably explained it poorly.
Those counts appear to be correct, however, `remove` is **always** called whereas `update` is called until a feature-gate is activated.

`remove` has 2 paths which is to remove uncommitted txs, or to remove the entire batch due to failed recording.
However, in both cases we loop over the transactions; I added in a count of the number of `Ok(cost)` in those loops (since those are what we added to cost-tracker) and remove the in-flight transactions for that count.

So in summary:
1. Batch adds all Ok(cost) to cost-tracker. Adds `N` to in-flight count.
2. Batch gets executed, try record, commit
3. Batch has costs updated (in front of feature-gate)
4. Batch has costs removed if not committed. Removes `N` from in-flight count, where `N` is the number of `Ok(cost)` **not** the number of committed transactions